src/govuk/components/character-count/character-count.unit.test.mjs

```diff
@@ -0,0 +1,118 @@
+        { number: 10, type: 'words', expected: 'You have 10 words remaining' },
+        { number: -1, type: 'words', expected: 'You have 1 word too many' },
+        { number: -10, type: 'words', expected: 'You have 10 words too many' },
+        { number: 0, type: 'words', expected: 'You have 0 words remaining' }
+      ]
+      it.each(cases)(
+        'picks the relevant translation for $number $type',
+        function test ({ number, type, expected }) {
+          expect(component.formatCountMessage(number, type)).toEqual(expected)
+        }
+      )
+
+      it('formats the number inserted in the message', () => {
+        expect(component.formatCountMessage(10000, 'words')).toEqual('You have 10,000 words remaining')
+        expect(component.formatCountMessage(-10000, 'words')).toEqual('You have 10,000 words too many')
+      })
+    })
+
+    describe('i18n', () => {
+      describe('JavaScript configuration', () => {
+        it('overrides the default translation keys', () => {
+          const component = new CharacterCount(document.createElement('div'), {
+            i18n: { charactersUnderLimitOne: 'Custom text. Count: %{count}' },
+            'i18n.charactersOverLimitOther': 'Different custom text. Count: %{count}'
+          })
+
+          expect(component.formatCountMessage(1, 'characters')).toEqual('Custom text. Count: 1')
+          expect(component.formatCountMessage(-10, 'characters')).toEqual('Different custom text. Count: 10')
+          // Other keys remain untouched
+          expect(component.formatCountMessage(10, 'characters')).toEqual('You have 10 characters remaining')
+        })
+
+        it('uses specific keys for when limit is reached', () => {
+          const component = new CharacterCount(document.createElement('div'), {
+            i18n: { charactersAtLimit: 'Custom text.' },
+            'i18n.wordsAtLimit': 'Different custom text.'
+          })
+
+          expect(component.formatCountMessage(0, 'characters')).toEqual('Custom text.')
+          expect(component.formatCountMessage(0, 'words')).toEqual('Different custom text.')
+        })
+      })
+
+      describe('lang attribute configuration', () => {
+        it('overrides the locale when set on the element', () => {
+          const $div = document.createElement('div')
+          $div.setAttribute('lang', 'de')
+
+          const component = new CharacterCount($div)
+
+          expect(component.formatCountMessage(10000, 'words')).toEqual('You have 10.000 words remaining')
+        })
+
+        it('overrides the locale when set on an ancestor', () => {
+          const $parent = document.createElement('div')
+          $parent.setAttribute('lang', 'de')
+          const $div = document.createElement('div')
+          $parent.appendChild($div)
+
+          const component = new CharacterCount($div)
+
+          expect(component.formatCountMessage(10000, 'words')).toEqual('You have 10.000 words remaining')
```

Comment:
Super minor one, but any thoughts on logically grouping what the test is doing?

```mjs
const $parent = document.createElement('div')
const $div = document.createElement('div')

$parent.setAttribute('lang', 'de')
$parent.appendChild($div)

const component = new CharacterCount($div)
const countMessage = component.formatCountMessage(10000, 'words')

expect(countMessage).toEqual('You have 10.000 words remaining')
```

1. Create things
2. Customise things
3. Run some code
4. Run assertion

Helps with readability slightly, not a blocker though